sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala

```diff
@@ -1753,6 +1753,23 @@ class SessionCatalog(
     }
   }
 
+  /**
+   * List all registered functions in a database with the given pattern.
+   */
+  private def listRegisteredFunctions(db: String, pattern: String): Seq[FunctionIdentifier] = {
+    val functions = (functionRegistry.listFunction() ++ tableFunctionRegistry.listFunction())
+      .filter(_.database.forall(_ == db))
+    StringUtils.filterPattern(functions.map(_.unquotedString), pattern).map { f =>
```

Comment:
I think the pattern is to match the function name, not the qualified function identifier.
```suggestion
    StringUtils.filterPattern(functions.map(_.name), pattern).map { f =>
```

Reply:
Ah, we probably need to add a variant of `StringUtils.filterPattern` which returns matching indexes.

Reply:
https://github.com/apache/spark/pull/35056#issue-1090368357 if we still want to support the legacy `SHOW FUNCTIONS LIKE 'db.func'`, then we probably can't filter out the registered functions from another database. 

But `SHOW FUNCTIONS LIKE 'db.func'` can only show **registered** functions from another database, and is different from `SHOW FUNCTIONS IN db LIKE 'func'`.